### PR TITLE
ID constructors

### PR DESCRIPTION
**Pull request by peterbourgon**
Use constructor functions to produce IDs in reports/topologies/tests. Remove manual, error-prone, and sometimes arbitrary string concatenation.

This is broken up into individual commits so you can see the progression, and if you have specific questions about a choice. I'll of course squash & rebase before merging.

### CODE REVIEW

File: report/id.go [hunk 1]
```diff
@@ -1,29 +1,117 @@
...
+	if len(fields) != 2 {
+		//log.Printf("AddressIDAddresser: bad input %q", id)
+		return nil
 	}
-	return net.ParseIP(parts[1])
+	return net.ParseIP(fields[1])
+}
+
+// PanicIDAddresser will panic if it's ever called. It's used in topologies
+// where there are never and edges, and so it's nonsensical to try and extract
```
Comment (tomwilkie) [inline report/id.go]: s/and/any/

**Comment by tomwilkie** [comment]
Might it be possible to use these ID functions instead of trySplitAddr in mapping_functions.go?

**Comment by tomwilkie** [comment]
Otherwise, LGTM

**Comment by peterbourgon** [comment]
> Might it be possible to use these ID functions instead of trySplitAddr in mapping_functions.go?

I agree it's a bit hacky. trySplitAddr is basically ParseArbitraryNodeID, since its callsites (pseudo funcs) just have opaque node IDs and don't know what topology they come from. So without changing how pseudo funcs work, we can't make it much smarter. I'll add a comment to this effect.

**Pull request merged**